libr/bin/format/mach0/coresymbolication.c

```diff
@@ -0,0 +1,368 @@
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_symbols;
+		for (i = 0; i < hdr->n_symbols && cursor + R_CS_EL_SIZE_SYM <= end; i++) {
+			RCoreSymCacheElementSymbol *sym = &result->symbols[i];
+			sym->paddr = r_read_le32 (cursor);
+			sym->size = r_read_le32 (cursor + 0x4);
+			sym->unk1 = r_read_le32 (cursor + 0x8);
+			size_t name_off = r_read_le32 (cursor + 0xc);
+			size_t mangled_name_off = r_read_le32 (cursor + 0x10);
+			sym->unk2 = (st32)r_read_le32 (cursor + 0x14);
+			sym->name = str_dup_safe (b, cursor + name_off, end);
+			if (!sym->name) {
+				break;
+			}
+			sym->mangled_name = str_dup_safe (b, cursor + mangled_name_off, end);
+			if (!sym->mangled_name) {
+				break;
+			}
+			cursor += R_CS_EL_SIZE_SYM;
+		}
+	}
+	const ut64 start_of_lined_symbols = start_of_symbols + (ut64)hdr->n_symbols * R_CS_EL_SIZE_SYM;
+	if (hdr->n_lined_symbols) {
+		result->lined_symbols = R_NEWS0 (RCoreSymCacheElementLinedSymbol, hdr->n_lined_symbols);
+		if (!result->lined_symbols) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_lined_symbols;
+		for (i = 0; i < hdr->n_lined_symbols && cursor + R_CS_EL_SIZE_LSYM <= end; i++) {
+			RCoreSymCacheElementLinedSymbol *lsym = &result->lined_symbols[i];
+			lsym->sym.paddr = r_read_le32 (cursor);
+			lsym->sym.size = r_read_le32 (cursor + 0x4);
+			lsym->sym.unk1 = r_read_le32 (cursor + 0x8);
+			size_t name_off = r_read_le32 (cursor + 0xc);
+			size_t mangled_name_off = r_read_le32 (cursor + 0x10);
+			lsym->sym.unk2 = (st32)r_read_le32 (cursor + 0x14);
+			size_t file_name_off = r_read_le32 (cursor + 0x18);
+			lsym->flc.line = r_read_le32 (cursor + 0x1c);
+			lsym->flc.col = r_read_le32 (cursor + 0x20);
+			lsym->sym.name = str_dup_safe (b, cursor + name_off, end);
+			if (!lsym->sym.name) {
+				break;
+			}
+			lsym->sym.mangled_name = str_dup_safe (b, cursor + mangled_name_off, end);
+			if (!lsym->sym.mangled_name) {
+				break;
+			}
+			lsym->flc.file = str_dup_safe (b, cursor + file_name_off, end);
+			if (!lsym->flc.file) {
+				break;
+			}
+			cursor += R_CS_EL_SIZE_LSYM;
+			meta_add_fileline (bf, r_coresym_cache_element_pa2va (result, lsym->sym.paddr), lsym->sym.size, &lsym->flc);
+		}
+	}
+	const ut64 start_of_line_info = start_of_lined_symbols + (ut64)hdr->n_lined_symbols * R_CS_EL_SIZE_LSYM;
+	if (hdr->n_line_info) {
+		result->line_info = R_NEWS0 (RCoreSymCacheElementLineInfo, hdr->n_line_info);
+		if (!result->line_info) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_line_info;
+		for (i = 0; i < hdr->n_line_info && cursor + R_CS_EL_SIZE_LINFO <= end; i++) {
+			RCoreSymCacheElementLineInfo *info = &result->line_info[i];
+			info->paddr = r_read_le32 (cursor);
+			info->size = r_read_le32 (cursor + 4);
+			size_t file_name_off = r_read_le32 (cursor + 8);
+			info->flc.line = r_read_le32 (cursor + 0xc);
+			info->flc.col = r_read_le32 (cursor + 0x10);
+			info->flc.file = str_dup_safe (b, cursor + file_name_off, end);
+			if (!info->flc.file) {
```

Comment:
if we break before the end, the result->line_info will not contain initialized items beyond the break point. we may want to continue here i think